File: csharp/LeetCode/001TwoSum.cs
using System;
using System.Collections.Generic;

namespace LeetCode
{
    // Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
    public class TwoSum
    {
        public int[] GetIndices(int[] nums, int target)
        {
            {
                var dic = new Dictionary<int, int>();

                for (var i = 0; i < nums.Length; i++)
                {
                    var complementary = target - nums[i];
                    if (dic.TryGetValue(complementary, out var index))
                    {
                        return new[] { i, index };
                    }

                    if (!dic.ContainsKey(nums[i]))
                    {
                        dic.Add(nums[i], i);
                    }
                }

                throw new NotImplementedException();
            }
        }
    }
}

File: rust/src/_001_two_sum.rs
use std::collections::HashMap;

// Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let mut scores: HashMap<i32, usize> = HashMap::new();
    
    for (i, num) in nums.into_iter().enumerate() {
        let diff = target - num;
        let value = scores.get(&diff);

        match value {
            Some(x) => {
                return vec![i.try_into().unwrap(), (*x).try_into().unwrap()];
            }
            None => {
                scores.insert(num, i);
            }
        }
    }

    panic!("Invalid input");
}
